worlds/pokemon_emerald/rules.py

```diff
@@ -15,78 +15,47 @@
-
-
-def _can_use_mach_bike(state: CollectionState, player: int) -> bool:
-    return state.has("Mach Bike", player)
-
-
-def _can_use_acro_bike(state: CollectionState, player: int) -> bool:
-    return state.has("Acro Bike", player)
-
-
-def _defeated_n_gym_leaders(state: CollectionState, player: int, n: int) -> bool:
-    num_gym_leaders_defeated = 0
-    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_ROXANNE", player) else 0
-    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_BRAWLY", player) else 0
-    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_WATTSON", player) else 0
-    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_FLANNERY", player) else 0
-    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_NORMAN", player) else 0
-    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_WINONA", player) else 0
-    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_TATE_AND_LIZA", player) else 0
-    num_gym_leaders_defeated += 1 if state.has("EVENT_DEFEAT_JUAN", player) else 0
-    return num_gym_leaders_defeated >= n
-
-
 # Rules are organized by town/route/dungeon and ordered approximately
 # by when you would first reach that place in a vanilla playthrough.
 def set_rules(world: PokemonEmeraldWorld) -> None:
     def can_cut(state: CollectionState):
-        return _can_cut(state, world.player)
+        return state.has("HM01 Cut", world.player) and state.has("Stone Badge", world.player)
 
     def can_surf(state: CollectionState):
-        return _can_surf(state, world.player)
+        return state.has("HM03 Surf", world.player) and state.has("Balance Badge", world.player)
 
     def can_strength(state: CollectionState):
-        return _can_strength(state, world.player)
+        return state.has("HM04 Strength", world.player) and state.has("Heat Badge", world.player)
 
     def can_flash(state: CollectionState):
-        return _can_flash(state, world.player)
+        return state.has("HM05 Flash", world.player) and state.has("Knuckle Badge", world.player)
 
     def can_rock_smash(state: CollectionState):
-        return _can_rock_smash(state, world.player)
+        return state.has("HM06 Rock Smash", world.player) and state.has("Dynamo Badge", world.player)
 
     def can_waterfall(state: CollectionState):
-        return _can_waterfall(state, world.player)
+        return state.has("HM07 Waterfall", world.player) and state.has("Rain Badge", world.player)
 
     def can_dive(state: CollectionState):
-        return _can_dive(state, world.player)
+        return state.has("HM08 Dive", world.player) and state.has("Mind Badge", world.player)
+
+    def has_acro_bike(state: CollectionState):
+        return state.has("Acro Bike", world.player)
+
+    def has_mach_bike(state: CollectionState):
+        return state.has("Mach Bike", world.player)
+    
+    def defeated_n_gym_leaders(state: CollectionState, n: int) -> bool:
+        num_gym_leaders_defeated = 0
```

Comment:
booleans are a subclass of int in python so this can all be simplified to ```num_gym_leaders_defeated = sum([state.has(event, world.player) for event in gym_leader_events])```

Reply:
5b1f8f3 More pleasing to the eye. I always feel a _little_ iffy adding bools together though.

Reply:
another alternative would be do essentially what you had earlier, but with a loop. it would have the benefit of being able to short circuit so would probably be just a tiny bit faster than this method, but it wouldn't read as well imo